runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkStatefulDoFnFunction.java

```diff
@@ -93,6 +101,14 @@ public void reduce(
     final K key = currentValue.getValue().getKey();
 
     final InMemoryStateInternals<K> stateInternals = InMemoryStateInternals.forKey(key);
+
+    // Used with Batch, we know that all the data is available for this key. We can't use the
+    // timer manager from the context because it doesn't exist. So we create one and emulate the
+    // watermark, knowing that we have all data and it is in timestamp order.
```

Comment:
That last comment is not true anymore since we don't sort the data by timestamp. I'll remove it before merging, OK?